File: scala/day4_part1_2015.scala
import scala.io.Source
import java.security.MessageDigest

object AdventCoins {
  def main(args: Array[String]): Unit = {
    val secretKey = Source.fromFile("input.txt").getLines().next()
    val md = MessageDigest.getInstance("MD5")

    def md5(s: String): String = {
      md.digest(s.getBytes).map("%02x".format(_)).mkString
    }

    def mineAdventCoin(key: String): Int = {
      Stream.from(1).find(i => md5(key + i).startsWith("00000")).get
    }

    val result = mineAdventCoin(secretKey)
    println(result)
  }
}

File: day4_part1_2015.clj
(require '[clojure.java.io :as io]
         '[clojure.string :as str]
         '[clojure.pprint :refer [pprint]])

(defn md5 [s]
  (let [md5 (java.security.MessageDigest/getInstance "MD5")
        bytes (.digest md5 (.getBytes s "UTF-8"))]
    (->> bytes
         (map #(format "%02x" %))
         (apply str))))

(defn find-advent-coin [input]
  (loop [n 1]
    (if (.startsWith (md5 (str input n)) "00000")
      n
      (recur (inc n)))))

(let [input (slurp "input.txt")]
  (println (find-advent-coin input)))
